test/unit/test-chunk.js

```diff
@@ -482,6 +482,113 @@ describe('long tasks', () => {
   );
 });
 
+describe('isInputPending usage', () => {
+  describes.fakeWin(
+    'pending input breaks microtask loop to subsequent macrotask',
+    {
+      amp: false,
+    },
+    (env) => {
+      let subscriptions;
+      let progress;
+      let postMessageCalls;
+      let pendingInput;
+
+      function complete(str, simulatePendingInputAfter) {
+        return function (unusedIdleDeadline) {
+          if (simulatePendingInputAfter) {
+            pendingInput = true;
+          }
+          progress += str;
+        };
+      }
+
+      function runSubs() {
+        subscriptions['message']
+          .slice()
+          .forEach((method) => method({data: 'amp-macro-task'}));
+      }
+
+      beforeEach(() => {
+        postMessageCalls = 0;
+        pendingInput = false;
+        subscriptions = {};
+
+        env.win.navigator.scheduling = {
+          isInputPending: function () {
+            if (pendingInput) {
+              pendingInput = false;
+              return true;
+            }
+
+            return false;
+          },
+        };
+
+        installDocService(env.win, /* isSingleDoc */ true);
+
+        env.win.addEventListener = function (type, handler) {
+          if (subscriptions[type] && !subscriptions[type].includes(handler)) {
+            subscriptions[type].push(handler);
+          } else {
+            subscriptions[type] = [handler];
+          }
+        };
+
+        env.win.postMessage = function (key) {
+          expect(key).to.equal('amp-macro-task');
+          postMessageCalls++;
+          runSubs();
+        };
+
+        progress = '';
+        chunkInstanceForTesting(
+          env.win.document.documentElement
+        ).macroAfterLongTask_ = true;
+      });
+
+      it('should not run macro tasks with invisible bodys', (done) => {
```

Comment:
FMI, could this also be worded as 'should not break out of microtask loop when body is invisible'